### ISSUE TITLE
`observer(component)` is not correctly recognized by typescript 1.8-beta

### ISSUE THREAD

**Issue by mweststrate** [mobxjs/mobx-react#29]
```
Hey. I'm trying out mobservable-react-typescript and I can't get the typescript compiler to accept "const TimerObserver = observer(Timer)" instead of @observer on Timer
I'm getting "Error:(22, 10) TS2604: JSX element type 'TimerObserver' does not have any construct or call signatures." when I try to use <TimerObserver />
If I cast it to any, the code works, just like the JS examples: const TimerObserver = observer(Timer) as any;

Hmm. Appears to be an issue with overload resolution in TypeScript. If I comment out all but the React.ComponentClass<P> overload it compiles.

TypeScript 1.8 beta(edited)
If I move the following overload to the top it works: export function observer<P>(clazz: 

React.ComponentClass<P>): React.ComponentClass<P>;
npm installed mobservable-react-typescript.. got mobservable 1.2.5 and mobservable-react 2.1.4
```

**Comment by mijay** [comment]
I have a similar issue. When I write the following (valid) code:
```
const ObservedMyComponent = observer(MyComponent);
createElement(ObservedMyComponent, { key: "12" });
```
I get the following error on the second line:
```
error TS2345: Argument of type 'ComponentClass<any> | void' is not assignable to parameter of type 'ComponentClass<any> | StatelessComponent<any>'.
  Type 'void' is not assignable to type 'ComponentClass<any> | StatelessComponent<any>'.
    Type 'void' is not assignable to type 'StatelessComponent<any>'.
```

After some juggling around, I've figured out that the problem is in the following line in your package typings:
```
export function observer<TFunction extends React.ComponentClass<any>>(target: TFunction): TFunction | void; // decorator signature
```
If you remove the `void` case, everything works as expected.

Typescript 1.7.5, React 0.14

**Comment by mweststrate** [comment]
I tried to reproduce this issue (see the above commit), however the above error didn't occur. 

Does the test https://github.com/mweststrate/mobservable-react/blob/master/test/ts/compile-ts.tsx properly reproduce the issue? Feel free to amend.

Test can simply be run by cloning the repo and running `tsc -p test/ts`

**Comment by mweststrate** [comment]
Maybe a different version of [react.d.ts](https://github.com/mweststrate/mobservable-react/blob/master/test/ts/react.d.ts)?

**Comment by mijay** [comment]
Didn't have a time to double check it yet, but your tests are missing the case where I got an error:
```
class T7 extends Component<{ pizza: number}, {}> {
	render() {
		return <div>{this.props.pizza}</div>;
	}
}
React.createElement(observer(T7), { pizza: "4 seasons" });
```

**Comment by mweststrate** [comment]
Tnx. That was the culprit indeed! Fixed and released as 2.1.5
